File: Leetcode/sbgo/Leetcode90-子集II.go
package sbgo

import "sort"

func subsetsWithDup(nums []int) (ans [][]int) {
	n := len(nums)
	path := []int{}
	used := make([]int, n)
	sort.Ints(nums)
	var dfs func(int, []int)
	dfs = func(index int, nums []int) {
		ans = append(ans, append([]int(nil), path...))
		for i := index; i < len(nums); i++ {
			if i > 0 && nums[i-1] == nums[i] && used[i-1] == 0 {
				continue
			}
			path = append(path, nums[i])
			used[i] = 1
			dfs(i+1, nums)
			path = path[:len(path)-1]
			used[i] = 0
		}
		return
	}
	dfs(0, nums)
	return
}

File: Leetcode/PY/Leetcode90-子集II.py
from typing import List

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        path,res= [],[]
        used = [0] * n
        nums.sort()
        def dfs(index):
            # 注意深拷贝
            res.append(path[:])
            for i in range(index,len(nums)):
                if i > 0 and nums[i - 1] == nums[i] and used[i - 1] == 0:
                    continue
                path.append(nums[i])
                used[i] =  1
                dfs(i + 1)
                used[i] = 0
                path.pop()

        dfs(0)

        return res